### ISSUE TITLE
Ambiguous params, differentiate Capture, Query String and Form Data

### ISSUE THREAD

**Issue by homam** [scotty-web/scotty#204]
How can I distinguish between params (Path Capture Params, Query String and Form Data) that have the same name?

For example, given this action handler 
```haskell
post "/:a" $ params >>= json
```

The following request

```bash
$ curl --data "a=fromData" ":3000/urlParam?a=queryString"
```

Returns

```javascript
[["a","urlParam"],["a","fromData"],["a","queryString"]]
```

How can I determine the source of `param "a"`?

Is there any particular reason that [here in code](https://github.com/scotty-web/scotty/blob/41ba9e41214ad0fdbf4a1d999e6b17e45f79c4f0/Web/Scotty/Route.hs#L189) we are concatenating all sources of params?

```haskell
parameters =  captures ++ map convert formparams ++ queryparams
```

**Comment by chris-martin** [comment]
The combining of params from all of these sources seems like a misfeature to me. In addition to this ambiguity, I also have an issue with the failure behavior. `param` produces a 500 HTTP status code when the param is missing - but this is only appropriate if the parameter was expected to come from path capture. If a query string or form parameter is required, the status code ought to be in the 400 range. Without any way to specify the intended source, correct behavior doesn't seem possible.

**Comment by ocramz** [comment]
The code is still there: https://github.com/scotty-web/scotty/blob/7e8739d7ac1e9c474229deca07a7ba4c857496b2/Web/Scotty/Route.hs#L199

**Comment by ocramz** [comment]
@chris-martin I've used your comment above as inspiration for the PR : https://github.com/scotty-web/scotty/pull/303

**Comment by ocramz** [comment]
Closed by #303
